### PR TITLE
Several improvements for lookup_embedder.penalty

### PR DESCRIPTION
**Pull request by samuelbroscheit**
- improve memory uage for unweighted penalty by using PyTorch builtin norm()

 - use call to embeddings in penalty to make sparse embeddings work, before this fix unewighted penalty for sparse embeddings did not work

### CODE REVIEW

File: kge/model/embedder/lookup_embedder.py [hunk 1]
```diff
@@ -105,13 +107,12 @@ def penalty(self, **kwargs) -> List[Tensor]:
             regularize_weight = self._get_regularize_weight()
             if not self.get_option("regularize_args.weighted"):
                 # unweighted Lp regularization
-                parameters = self.embeddings.weight
```
Comment (samuelbroscheit) [inline kge/model/embedder/lookup_embedder.py:108]: This made sparse embeddings break

File: kge/model/embedder/lookup_embedder.py [hunk 2]
```diff
@@ -86,7 +87,8 @@ def embed(self, indexes: Tensor) -> Tensor:
         return self._embed(self.embeddings(indexes.long()))
 
     def embed_all(self) -> Tensor:
-        return self._embed(self.embeddings.weight)
```
Comment (samuelbroscheit) [inline kge/model/embedder/lookup_embedder.py:89]: This caused that dropout was incorrectly not applied to score funcs that call embed_all

File: kge/model/embedder/lookup_embedder.py [hunk 3]
```diff
@@ -105,13 +107,12 @@ def penalty(self, **kwargs) -> List[Tensor]:
...
                 # unweighted Lp regularization
-                parameters = self.embeddings.weight
-                if p % 2 == 1:
-                    parameters = torch.abs(parameters)
+                self.all_indexes = self.all_indexes.to(self.embeddings.weight.device)
+                parameters = self.embeddings(self.all_indexes)
                 result += [
                     (
                         f"{self.configuration_key}.L{p}_penalty",
-                        (regularize_weight / p * (parameters ** p)).sum(),
```
Comment (samuelbroscheit) [inline kge/model/embedder/lookup_embedder.py:114]: This caused huge memory consumption for RESCAL

File: kge/model/embedder/lookup_embedder.py [hunk 4]
```diff
@@ -86,7 +87,8 @@ def embed(self, indexes: Tensor) -> Tensor:
         return self._embed(self.embeddings(indexes.long()))
 
     def embed_all(self) -> Tensor:
-        return self._embed(self.embeddings.weight)
+        self.all_indexes = self.all_indexes.to(self.embeddings.weight.device)
+        return self.embed(self.all_indexes)
```
Comment (rgemulla) [inline kge/model/embedder/lookup_embedder.py:91]: I think this would be handled much better (computation, storage) if a `None` argument was interpreted as "all_indexes" my `embed`.

File: kge/model/embedder/lookup_embedder.py [hunk 5]
```diff
@@ -105,13 +107,12 @@ def penalty(self, **kwargs) -> List[Tensor]:
             regularize_weight = self._get_regularize_weight()
             if not self.get_option("regularize_args.weighted"):
                 # unweighted Lp regularization
-                parameters = self.embeddings.weight
-                if p % 2 == 1:
-                    parameters = torch.abs(parameters)
+                self.all_indexes = self.all_indexes.to(self.embeddings.weight.device)
+                parameters = self.embeddings(self.all_indexes)
```
Comment (rgemulla) [inline kge/model/embedder/lookup_embedder.py:111]: Likewise.

Comment (samuelbroscheit) [reply kge/model/embedder/lookup_embedder.py:111]: Here we do not call embed

File: kge/model/embedder/lookup_embedder.py (hunk 4 above)
Comment (samuelbroscheit) [reply kge/model/embedder/lookup_embedder.py:91]: I don't understand how this would improve computation and storage, but I can shift it over to embed. 

so basically you propose:

```
    def embed(self, indexes: Tensor) -> Tensor:
        if indexes is None:
            self.all_indexes = self.all_indexes.to(self.embeddings.weight.device)
            indexes = self.all_indexes 
        return self._embed(self.embeddings(indexes.long()))

```

File: kge/model/embedder/lookup_embedder.py (hunk 5 above)
Comment (rgemulla) [reply kge/model/embedder/lookup_embedder.py:111]: but embeddings

File: kge/model/embedder/lookup_embedder.py (hunk 4 above)
Comment (rgemulla) [reply kge/model/embedder/lookup_embedder.py:91]: No. Never use all_indexes (do not even create it). Change the API such that indexes=None means all indexes. In this case, we should return self.embeddings.weight. Or am I missing something? It's a waste of space to store a vector of 1...num_embeddings on the GPU.

Comment (samuelbroscheit) [reply kge/model/embedder/lookup_embedder.py:91]: Accessing self.embeddings.weight was the problem that is being fixed here. The point of the change is that we don't get sparse gradients when we directly access self.embeddings.weight. Please note that sparse gradients is not achieved by self.embeddings.weight.to_sparse(), only the gradients are sparse. 

However, this is fixing the case of using SparseAdam+unweighted penalty, SparseAdam+1vsAll or SparseAdam+KvsAll. One could argue that those are non-sensical anyway because sparse is most likely used when we do not want to access the whole entity embedding table, i.e., for sampled training strategies and weighted penalties. We could just opt to not support that combination and not waste those 7MB / 1M entities of long indices.

Comment (rgemulla) [reply kge/model/embedder/lookup_embedder.py:91]: I think we should either use `.weight` or `.embeddings()` (i.e., forward), but not mix both. 

I am fine to consistently use `.embeddings()`. But then write `torch.arange(..., device=..., dtype=torch.long)` to get all indexes. I don't think there is any benefit of caching it.

Agreed?

Comment (samuelbroscheit) [reply kge/model/embedder/lookup_embedder.py:91]: Ok, I thought allocating the exact same tensor constantly is more wasteful, also, with having the GPU-memory fragmentation issue in mind ... . But I'll trust your judgment.

Comment (rgemulla) [reply kge/model/embedder/lookup_embedder.py:91]: Let's keep it simple and only add caching etc. when we run into problems.

File: kge/model/embedder/lookup_embedder.py [hunk 6]
```diff
@@ -84,11 +83,20 @@ def _embed(self, embeddings: Tensor) -> Tensor:
...
+            return self._embed(self._get_all_parameters())
+        else:
+            return self._embed(self.embeddings(indexes.long()))
 
     def embed_all(self) -> Tensor:
-        self.all_indexes = self.all_indexes.to(self.embeddings.weight.device)
-        return self.embed(self.all_indexes)
+        return self.embed(None)
+
+    def _get_all_parameters(self) -> Tensor:
```
Comment (rgemulla) [inline kge/model/embedder/lookup_embedder.py:94]: Names are quite confusing (already before this PR). I suggest:
- Rename `self.embeddings` to `self._embeddings`
- Rename this function to `self._embeddings_all` (for consistency with embed and embed_all)
- Rename `self._embed` to `self._postprocess`

**Comment by rgemulla** [comment]
Thanks!
